src/builder.ts

```diff
@@ -0,0 +1,132 @@
+/**
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+import {CpuProfile, CpuProfileNode} from './v8-types';
+import {perftools} from './profile';
+
+// TODO: CpuProfiler::sample_interval_ can be customized.. should query that
+const SAMPLE_INTERVAL = 1000;
+
+// A stack of function UIDs.
+type Stack = Array<number>;
+
+let strings = [''];
+let samples: Array<perftools.profiles.Sample> = [];
+
+let locations: Array<perftools.profiles.Location> = [];
+let locationMap: Map<number, perftools.profiles.Location>;
+
+let functions: Array<perftools.profiles.Function> = [];
+let functionMap: Map<number, perftools.profiles.Function>;
+
+function getStringIndex(str: string) {
+  let index = strings.indexOf(str);
+  if (index !== -1) {
+    return index;
+  }
+  index = strings.push(str);
+  return index - 1;
+}
+
+function getFunction(node: CpuProfileNode) {
+  let id = node.callUid;
+  if (functionMap.has(id)) {
+    // Map.get returns possibly undefined, but we know it is defined.
+    // TODO: figure out how to avoid the cast.
```

Comment:
I think the only way to avoid a cast presently is to `get` first, and then check that the return value isn't `undefined`.

Reply:
I'll leave it as a TODO for now.